services/src/kilda-core/kilda-persistence-neo4j/src/main/java/org/openkilda/persistence/repositories/impl/Neo4jFlowRepository.java

```diff
@@ -0,0 +1,191 @@
+                + "RETURN f.flowid", parameters).forEach(flowIds::add);
+        return flowIds;
+    }
+
+    @Override
+    public Collection<String> findActiveFlowIdsWithPortInPath(SwitchId switchId, int port) {
+        Map<String, Object> parameters = ImmutableMap.of(
+                SWITCH_ID_PROPERTY_NAME, switchId.toString(),
+                "port", port,
+                FLOW_STATUS_PROPERTY_NAME, flowStatusConverter.toGraphProperty(FlowStatus.UP));
+
+        Set<String> flowIds = new HashSet<>();
+        // Treat empty status as UP to support old storage schema.
+        getSession().query(String.class, "MATCH (src:switch)-[f:flow]->(dst:switch) "
+                + "WHERE (src.name=$switch_id AND f.src_port=$port "
+                + " OR dst.name=$switch_id AND f.dst_port=$port) "
+                + " AND (f.status=$flow_status OR f.status IS NULL)"
+                + "RETURN f.flowid", parameters).forEach(flowIds::add);
+
+        getSession().query(String.class, "MATCH (src:switch)-[fs:flow_segment]->(dst:switch) "
+                + "WHERE (src.name=$switch_id AND fs.src_port=$port "
+                + " OR dst.name=$switch_id AND fs.dst_port=$port) "
+                + "WITH fs "
+                + "MATCH ()-[f:flow]->() "
+                + "WHERE fs.flowid = f.flowid AND (f.status=$flow_status OR f.status IS NULL)"
+                + "RETURN f.flowid", parameters).forEach(flowIds::add);
+        return flowIds;
+    }
+
+    @Override
+    public Collection<String> findDownFlowIds() {
+        Map<String, Object> parameters = ImmutableMap.of(
+                FLOW_STATUS_PROPERTY_NAME, flowStatusConverter.toGraphProperty(FlowStatus.DOWN));
+
+        Set<String> flowIds = new HashSet<>();
+        getSession().query(String.class,
+                "MATCH ()-[f:flow{status: {flow_status}}]->() RETURN f.flowid", parameters).forEach(flowIds::add);
+        return flowIds;
+    }
+
+    @Override
+    public void createOrUpdate(Flow flow) {
+        Map<SwitchId, Switch> lockedSwitches = lockSwitches(flow.getSrcSwitch(), flow.getDestSwitch());
+
+        flow.setSrcSwitch(lockedSwitches.get(flow.getSrcSwitch().getSwitchId()));
+        flow.setDestSwitch(lockedSwitches.get(flow.getDestSwitch().getSwitchId()));
+
+        super.createOrUpdate(flow);
+    }
+
+    @Override
+    public void createOrUpdate(FlowPair flowPair) {
+        createOrUpdate(flowPair.getForward());
+        createOrUpdate(flowPair.getReverse());
+    }
+
+    @Override
+    public void updateStatus(String flowId, FlowStatus status) {
+        Map<String, Object> parameters = ImmutableMap.of(
+                FLOW_ID_PROPERTY_NAME, flowId,
+                FLOW_STATUS_PROPERTY_NAME, flowStatusConverter.toGraphProperty(status));
+
+        String query = "MATCH ()-[f:flow{flowid: {flowid}}]->() SET f.status=$flow_status RETURN id(f)";
+
+        Iterable<Long> flowRelationIds = getSession().query(Long.class, query, parameters);
```

Comment:
This is incorrect for ORM approach. We must not perform direct "update" queries. Because corresponding object can be loaded and used by some other piece of code. And it will use stale object in this case.

Call to the `session.detach*Entity` make situation even worse - we "extract" object from session and if it is stored/used by some other code - it will lead to lost update or to creation of duplicated object.

For ORM approach we must load mapped object, change object and save it.

We can use update queries to change multiple object ... but extremely carefully - we must know that we do not use corresponding objects anywhere else.

Reply:
Agree. Replaced with fetching Flow entity and updating the status.